File: harness/semgrep/fixtures/compare-catch/verify.good.py
# GOOD: constant-time compare via hmac.compare_digest, and except fails closed.

import hmac
import hashlib


def check_mac(payload: bytes, secret_key: bytes, provided_mac: str) -> bool:
    expected_mac = hmac.new(secret_key, payload, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected_mac, provided_mac)


def verify_token(token: str) -> bool:
    try:
        assert_signature_valid(token)
        return True
    except Exception:
        return False

File: harness/semgrep/fixtures/compare-catch/verify.good.go
package verify

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
)

// GOOD: constant-time compare via subtle.ConstantTimeCompare, and errors propagate.

func checkMac(payload, secretKey []byte, providedMac []byte) bool {
	mac := hmac.New(sha256.New, secretKey)
	mac.Write(payload)
	expectedMac := mac.Sum(nil)
	return subtle.ConstantTimeCompare(expectedMac, providedMac) == 1
}

func verifyToken(token string) (bool, error) {
	err := assertSignatureValid(token)
	if err != nil {
		return false, err
	}
	return true, nil
}
